### PR TITLE
CDAP-8918 Fix schedule deletion during app re-deploy

### PR DESCRIPTION
**Pull request by rohitsinha54**
Issue: https://issues.cask.co/browse/CDAP-8918
Build: http://builds.cask.co/browse/CDAP-RUT849-1

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java [hunk 1]
```diff
@@ -315,7 +315,6 @@ public void apply() throws Exception {
                                                       AbstractSchedulerService.scheduleIdFor(programId, programType,
                                                                                              scheduleName)));
           table.put(rowKey, columns, values);
-          LOG.debug("Updated schedule {} with columns {}, values {}", scheduleName, columns, values);
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:318]: Unintended removal?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    // drop deleted schedules
+    Map<String, ScheduleSpecification> existingSchedulesMap = input.getExistingAppSpec() != null ?
+      input.getExistingAppSpec().getSchedules() : ImmutableMap.<String, ScheduleSpecification>of();
+    MapDifference<String, ScheduleSpecification> mapDiff = Maps.difference(existingSchedulesMap,
+                                                                           input.getSpecification().getSchedules());
+    for (Map.Entry<String, ScheduleSpecification> entry : mapDiff.entriesOnlyOnLeft().entrySet()) {
+      // delete schedules that existed in the old app spec, but don't anymore
+      ScheduleSpecification scheduleSpec = entry.getValue();
+      ProgramType programType = ProgramType.valueOfSchedulableType(scheduleSpec.getProgram().getProgramType());
+      LOG.debug("Deleting schedule {} with specification {}", entry.getKey(), scheduleSpec);
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java:59]: Would be good to add the program id in the log too

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    // CreateScheduleStage with new type and spec. Its required to delete it here because in next stage the
+    // application will be registered updating the schedule spec. See CDAP-8918 for details.
+    for (Map.Entry<String, MapDifference.ValueDifference<ScheduleSpecification>> entry :
+      mapDiff.entriesDiffering().entrySet()) {
+      ScheduleSpecification newScheduleSpec = entry.getValue().rightValue();
+      ScheduleSpecification oldScheduleSpec = entry.getValue().leftValue();
+      if (ScheduleType.fromSchedule(newScheduleSpec.getSchedule()) !=
+        ScheduleType.fromSchedule(oldScheduleSpec.getSchedule())) {
+        LOG.debug("Deleting existing schedule {} with specification {}. It will be created with specification {} " +
+                    "as they differ in schedule type.", entry.getKey(), oldScheduleSpec, newScheduleSpec);
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java:75]: Same here about adding program id in the log message

**Review by poornachandra — COMMENTED** [review]
@rohitsinha54 Code changes look good. Few minor comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:318]: I removed it on purpose as this debug line looks not helpful. It prints bytes and pollutes the logs

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java:59]: I am printing the schedule spec which will have the program info correct ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java (hunk 3 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java:75]: Same as above.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java [hunk 4]
```diff
@@ -16,26 +16,34 @@
...
+ * specified by the application. If the schedules already exist, it will update them. For schedule deletion see
+ * {@link DeleteScheduleStage}. They are broken into separate stages because we want schedule creation and update to
+ * happen after the app is registered in the app store as it might be possible that a schedule gets trigged instantly
+ * after being added when the app is still to be registered. See CDAP-8918 for details.
  */
 public class CreateSchedulesStage extends AbstractStage<ApplicationWithPrograms> {
 
-  private final Scheduler scheduler;
+  private static final Logger LOG = LoggerFactory.getLogger(CreateSchedulesStage.class);
+  protected final Scheduler scheduler;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: Why protected field? Who is extending this class?

Comment (rohitsinha54) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: Oops it was from abstract class which we removed. Will make it private.

**Comment by rohitsinha54** [comment]
@poornachandra @chtyim : Is this good to merge ?

**Comment by poornachandra** [comment]
LGTM

**Review by poornachandra — APPROVED** [review]

**Pull request merged**